cpp/include/cudf/detail/null_mask.hpp

```diff
@@ -87,6 +87,21 @@ rmm::device_buffer copy_bitmask(
   rmm::cuda_stream_view stream,
   rmm::mr::device_memory_resource *mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @copydoc bitmask_binop(Binop op, std::vector<bitmask_type const*>, std::vector<size_type> const&,
+ * size_type, rmm::mr::device_memory_resource *)
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ */
+template <typename Binop>
```

Comment:
These don't look right. If it's a template, then the definition of the function needs to be in the header.

Reply:
Technically you can declare templated functions in a header and define them in a .cpp file. As long as all the possible templated functions are defined somewhere the compiler will resolve all the function calls at link time. This does not apply for us though for inlined device code.

Reply:
Right, but the problem is the "all possible instantiations". If I were to include this header in some other TU and pass in a `Binop` that wasn't instantiated already in `null_mask.cu`, then it wouldn't work.

Reply:
See https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl

Reply:
Options:
1. Don't expose these functions in a header at all, just put them in an anon namespace inside `null_mask.cu`
2. If they will be exposed in a header, then their definitions need to be available.